### PR TITLE
fix coercion of null for decimal math in binary_rules

### LINKED ISSUE #3479: coercion between decimal and other types lacking, compared to other numeric types
**Describe the bug**

**To Reproduce**
From test sql::expr::binary_mathematical_operator_with_null_lt 
```sql
❯ select null + 1::decimal;
Plan("'Null + Decimal128(38, 10)' can't be evaluated because there isn't a common type to coerce the types to")
```

or (from the "test_string_concat_operator" test):
```sql
❯ select 'a' || 1.1::decimal;
Plan("'Utf8 || Decimal128(38, 10)' can't be evaluated because there isn't a common type to coerce the types to")
```

**Expected behavior**

It should work like floats do:
```sql
❯ select null + 1::float;
+-----------------+
| NULL + Int64(1) |
+-----------------+
|                 |
+-----------------+
1 row in set. Query took 0.005 seconds.
```

```sql
❯ select 'a' || 1.1;
+---------------------------+
| Utf8("a") || Float64(1.1) |
+---------------------------+
| a1.1                      |
+---------------------------+
1 row in set. Query took 0.003 seconds.
```

**Additional context**

### PR DESCRIPTION
**Pull request by kmitchener**
# Which issue does this PR close?

Closes #3479 . This fix can close that issue -- there's another issue for adding support for casting utf8 to decimal.

 # Rationale for this change

Prior to this fix:
```sql
❯ select null + 1::decimal;
Plan("'Null + Decimal128(38, 10)' can't be evaluated because there isn't a common type to coerce the types to")
```

After this fix:
```sql
❯ select null + 1::decimal;
+-----------------+
| NULL + Int64(1) |
+-----------------+
|                 |
+-----------------+
1 row in set. Query took 0.020 seconds.
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/binary_rule.rs [hunk 1]
```diff
@@ -308,25 +308,18 @@ fn mathematics_numerical_coercion(
...
-            match converted_decimal_type {
-                None => None,
-                Some(right_decimal_type) => coercion_decimal_mathematics_type(
-                    mathematics_op,
-                    lhs_type,
-                    &right_decimal_type,
-                ),
-            }
+        (Null, dec_type @ Decimal128(_, _)) | (dec_type @ Decimal128(_, _), Null) => {
+            Some(dec_type.clone())
```
Comment (kmitchener) [inline datafusion/expr/src/binary_rule.rs:312]: this is the actual fix, the rest is some refactoring

**Comment by liukun4515** [comment]
> # Which issue does this PR close?
> It fixes part of #3479 Closes #.
> 
> # Rationale for this change
> Prior to this fix:
> 
> ```sql
> ❯ select null + 1::decimal;
> Plan("'Null + Decimal128(38, 10)' can't be evaluated because there isn't a common type to coerce the types to")
> ```
> 
> After this fix:
> 
> ```sql
> ❯ select null + 1::decimal;
> +-----------------+
> | NULL + Int64(1) |
> +-----------------+
> |                 |
> +-----------------+
> 1 row in set. Query took 0.020 seconds.
> ```
> 
> # What changes are included in this PR?
> # Are there any user-facing changes?

why the result is `null+int64` not the decimal?

**Comment by kmitchener** [comment]
@liukun4515 it's just the naming of the expression. the actual datatype is Decimal.

```sql
❯ select arrow_typeof(null + 1::decimal);
+------------------------------+
| arrowtypeof(NULL + Int64(1)) |
+------------------------------+
| Decimal128(38, 10)           |
+------------------------------+
1 row in set. Query took 0.045 seconds.
```

Naming of resulting columns after casting is kinda weird, but unrelated to this PR.
```sql
❯ select 1::decimal;
+--------------+
| Int64(1)     |
+--------------+
| 1.0000000000 |
+--------------+
1 row in set. Query took 0.008 seconds.
❯ select 1::float;
+----------+
| Int64(1) |
+----------+
| 1        |
+----------+
1 row in set. Query took 0.005 seconds.
❯ select 1::text;
+----------+
| Int64(1) |
+----------+
| 1        |
+----------+
1 row in set. Query took 0.005 seconds.
```

File: datafusion/expr/src/binary_rule.rs [hunk 2]
```diff
@@ -624,19 +627,12 @@ fn eq_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
...
-            if can_cast_types(&DataType::Null, rhs_type) {
-                Some(rhs_type.clone())
-            } else {
-                None
-            }
-        }
-        (_, DataType::Null) => {
-            if can_cast_types(&DataType::Null, lhs_type) {
-                Some(lhs_type.clone())
+        (DataType::Null, other_type) | (other_type, DataType::Null) => {
```
Comment (alamb) [inline datafusion/expr/src/binary_rule.rs:633]: this is a nice cleanup

File: datafusion/expr/src/binary_rule.rs (hunk 1 above)
Comment (alamb) [reply datafusion/expr/src/binary_rule.rs:312]: Could the same coercion (Null to that type) be applied for all the other numeric types as well? 

In other words, I wonder if we need to special case Decimal128 ? Is it to skip the special decimal coercion logic below?

**Review by alamb — APPROVED** [review]
LGTM -- thanks @kmitchener

File: datafusion/expr/src/binary_rule.rs (hunk 1 above)
Comment (kmitchener) [reply datafusion/expr/src/binary_rule.rs:312]: Yes, it's basically to shortcut that next match arm for figuring out the Decimal's precision and scale, which depends on the datatype being math'd to the decimal ... none of that matters if it's a null, you just need to return the same datatype. Back further up the chain, the null gets cast to this datatype.

The other number types don't have the rules that Decimals do, so you'll see in the same match they just return their own datatype, no complexity.

**Comment by liukun4515** [comment]
> @liukun4515 it's just the naming of the expression. the actual datatype is Decimal.
> 
> ```sql
> ❯ select arrow_typeof(null + 1::decimal);
> +------------------------------+
> | arrowtypeof(NULL + Int64(1)) |
> +------------------------------+
> | Decimal128(38, 10)           |
> +------------------------------+
> 1 row in set. Query took 0.045 seconds.
> ```
> 
> Naming of resulting columns after casting is kinda weird, but unrelated to this PR.
> 
> ```sql
> ❯ select 1::decimal;
> +--------------+
> | Int64(1)     |
> +--------------+
> | 1.0000000000 |
> +--------------+
> 1 row in set. Query took 0.008 seconds.
> ❯ select 1::float;
> +----------+
> | Int64(1) |
> +----------+
> | 1        |
> +----------+
> 1 row in set. Query took 0.005 seconds.
> ❯ select 1::text;
> +----------+
> | Int64(1) |
> +----------+
> | 1        |
> +----------+
> 1 row in set. Query took 0.005 seconds.
> ```

But this make me confused.
I think it's issue related with https://github.com/apache/arrow-datafusion/issues/3568

cc @alamb  Do you have any comments about this.

this fix looks good to me expect this  header.

**Comment by alamb** [comment]
> cc @alamb Do you have any comments about this.


I agree the header name for constants looks non ideal. I seem to remember a discussion we had somewhere else to maybe make the names of constants different but I couldn't find it

**Comment by liukun4515** [comment]
> I seem to remember a discussion we had somewhere else to maybe make the names of constants different but I couldn't find it

Maybe this can help us. https://github.com/apache/arrow-datafusion/issues/3568#issuecomment-1253868778

**Pull request merged**